high_performance_computing/hpc_openmp/02_intro_openmp.md

```diff
@@ -0,0 +1,146 @@
+OpenMP consists of three key components that enable parallel programming using threads:
+
+- **Compiler directives:** OpenMP makes use of special code markers known as *compiler directives* to indicate to the compiler when and how to parallelise various sections of code. These directives are prefixed with `#pragma omp`, and mark sections of code to be executed concurrently by multiple threads.
+- **Runtime Library Routines:** These are predefined functions provided by the OpenMP runtime library. They allow you to control the behavior of threads, manage synchronization, and handle parallel execution. For example, we can use the function `omp_get_thread_num()` to obtain the unique identifier of the calling thread.
+- **Environment Variables:** These are settings that can be adjusted to influence the behavior of the OpenMP runtime. They provide a way to fine-tune the parallel execution of your program. Setting OpenMP environment variables is typically done similarly to other environment variables for your system. For instance, you can adjust the number of threads to use for a program you are about to execute by specifying the value in the `OMP_NUM_THREADS` environment variable.
+
+
+## Running a Code with OpenMP
+
+Before we delve into specifics of writing code that uses OpenMP, let's first look at how we compile and run an example "Hello World!" OpenMP program that prints this to the console.
+
+ First, log into whichever HPC system you have access to - this could be a group server, or university- or national-level cluster (e.g. Iridis or DiRAC). Alternatively, you may have OpenMP installed on your local machine.
+ 
+Save the following code in `hello_world_omp.c`:
+
+~~~c
+#include <stdio.h>
+#include <omp.h>
+
+int main() {
+    #pragma omp parallel
+    {
+        printf("Hello World!\n");
+    }
+}
+~~~
+
+You'll likely want to compile it using a standard compiler such as `gcc`, although this may depend on your system. To enable the creation of multi-threaded code based on OpenMP directives, pass the `-fopenmp` flag to the compiler. This flag indicates that you're compiling an OpenMP program:
+
+~~~bash
+gcc hello_world_omp.c -o hello_world_omp -fopenmp
+~~~
+
+Before we run the code we also need to indicate how many threads we wish the program to use.
+One way to do this is to specify this using the `OMP_NUM_THREADS` environment variable, e.g.
+
+~~~bash
+export OMP_NUM_THREADS=4
+~~~
+
+Now you can run it just like any other program using the following command:
+
+~~~bash
+./hello_world_omp
+~~~
+
+When you execute the OpenMP program,
+it will display 'Hello World!' multiple times according to the value we entered in `OMP_NUM_THREADS`,
+with each thread in the parallel region executing the `printf` statement concurrently:
+
+~~~
+Hello World!
+Hello World!
+Hello World!
+Hello World!
+~~~
+
+## OpenMP vs. Low-Level Threading APIs (POSIX Threads)
+
+When it comes to parallel programming with threads, there are two main ways to tackle it: the 
```

Comment:
there are many other libraries for parallel programming that are built on threading (e.g. intell TBB)

Reply:
Addressed via its removal (but it might be good to eventually have a section discussing alternatives at this level).